### PR TITLE
Encode ASTC gpu compressed textures

### PR DESCRIPTION
**Pull request by fire**
Proposal from reduz for astc support.

See discussion from https://github.com/godotengine/godot/pull/65376

WORK IN PROGRESS

- [x] getting the shape of the pr
- [x] handle hdr
- [x] Will not interpolate to multiples of 4 or 8.
- [x] Tested astc
- [x] Windows support

Core of the work is in modules/astcenc/image_compress_astcenc.cpp.

### CODE REVIEW

**Comment by fire** [comment]
@reduz Do you prefer for me to test for the range of colors outside 0 and 1 or add a flag for hdr?

**Comment by RevoluPowered** [comment]
Going to test this on mac, I have a game I can run this on. Interested to see ASTC on our project.

Will let you know the results if there are bugs etc :)

Will try intel macbook i9 with radeon pro, latest macos etc.

File: modules/astcenc/image_compress_astcenc.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	if (img_format >= Image::FORMAT_DXT1) {
+		return; // Do not compress, already compressed.
+	}
+	if (img_format > Image::FORMAT_RGBA8) {
+		// TODO: we should be able to handle FORMAT_RGBA4444 and FORMAT_RGBA5551 eventually
+		return;
+	}
+
+	// Use RGBA8 to convert.
+	if (img_format != Image::FORMAT_RGBA8) {
```
Comment (fire) [inline modules/astcenc/image_compress_astcenc.cpp:50]: RGBA8 or RGBAH or RGBAF

File: modules/astcenc/image_compress_astcenc.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+	astcenc_context *context;
+	status = astcenc_context_alloc(&config, thread_count, &context);
+	ERR_FAIL_COND_MSG(status != ASTCENC_SUCCESS, vformat("ASTC context allocation failed. %s", astcenc_get_error_string(status)));
+
+	astcenc_image image;
+	image.dim_x = width;
+	image.dim_y = height;
+	image.dim_z = 1;
+	image.data_type = ASTCENC_TYPE_U8;
```
Comment (fire) [inline modules/astcenc/image_compress_astcenc.cpp:95]: Switch on 8 bit, 16 bit float or 32 bit float.

**Comment by fire** [comment]
Comments on where the last work:

> The code compiles, there a really small part where you convert byte buffer to astc input arguments. I'm missing that but the astc code should be in the pr.
> I think the 10-30 lines of code to convert Image byte array to ASTC input is not that hard
> tl;dr todo is take Image packed buffer. generate an enum for rgba or rgbf and use the astc maintainer's strategy for SRGB vs linear (in comment)
> then take enum and the byte buffer and send it to astc
> then cast the byte array to an image format of godot

**Comment by fire** [comment]
Closing this pr but will leave the branch for posterity until https://github.com/godotengine/godot/pull/70363 is merged and the feature add astc import is done.
